webhook/url_notifier.go

```diff
@@ -0,0 +1,185 @@
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/frostbyte73/core"
+	"github.com/gammazero/deque"
+	"github.com/hashicorp/go-retryablehttp"
+	"go.uber.org/atomic"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	"github.com/livekit/protocol/auth"
+	"github.com/livekit/protocol/livekit"
+	"github.com/livekit/protocol/logger"
+)
+
+type URLNotifierParams struct {
+	Logger    logger.Logger
+	QueueSize int
+	URL       string
+	APIKey    string
+	APISecret string
+}
+
+const defaultQueueSize = 100
+
+var ErrNotifierStopped = errors.New("notifier has already been stopped")
+
+// URLNotifier is a QueuedNotifier that sends a POST request to a Webhook URL.
+// It will retry on failure, and will drop events if notification fall too far behind
+type URLNotifier struct {
+	params    URLNotifierParams
+	queue     *deque.Deque[*livekit.WebhookEvent]
+	client    *retryablehttp.Client
+	draining  atomic.Bool
+	dropped   atomic.Int32
+	fuse      core.Fuse
+	mu        sync.RWMutex
+	done      chan struct{}
+	jobSignal chan struct{}
+}
+
+func NewURLNotifier(params URLNotifierParams) *URLNotifier {
+	if params.QueueSize == 0 {
+		params.QueueSize = defaultQueueSize
+	}
+	if params.Logger == nil {
+		params.Logger = logger.GetLogger()
+	}
+	return &URLNotifier{
+		params:    params,
+		client:    retryablehttp.NewClient(),
+		queue:     deque.New[*livekit.WebhookEvent](),
+		fuse:      core.NewFuse(),
+		jobSignal: make(chan struct{}, 1),
+	}
+}
+
+func (n *URLNotifier) Start() {
+	if n.done != nil {
+		return
+	}
+	n.done = make(chan struct{})
+	go n.worker()
+}
+
+func (n *URLNotifier) Stop(force bool) {
+	if force {
+		// triggers immediate closure
+		n.fuse.Break()
+	} else {
+		// closes after current queue is processed
+		n.draining.Store(true)
+	}
+
+	if !force {
+		// wait for current queue to be processed
+		<-n.done
+	}
+}
+
+func (n *URLNotifier) SetKeys(apiKey, apiSecret string) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	n.params.APIKey = apiKey
+	n.params.APISecret = apiSecret
+}
+
+func (n *URLNotifier) QueueNotify(event *livekit.WebhookEvent) error {
+	if n.draining.Load() || n.fuse.IsBroken() {
+		return ErrNotifierStopped
+	}
+	n.mu.Lock()
+	n.queue.PushBack(event)
+	if n.queue.Len() > n.params.QueueSize {
+		n.dropped.Inc()
+		n.queue.PopFront()
+	}
+	n.mu.Unlock()
+	select {
+	case n.jobSignal <- struct{}{}:
+	default:
+	}
+	return nil
+}
+
+func (n *URLNotifier) worker() {
+	defer close(n.done)
+
+	waitTicker := time.NewTicker(100 * time.Millisecond)
+	for !n.fuse.IsBroken() {
+		select {
+		case <-waitTicker.C:
+		case <-n.fuse.Watch():
+			return
+		case <-n.jobSignal:
+			n.processQueue()
+			if n.draining.Load() {
+				return
+			}
+		}
+	}
+}
+
+func (n *URLNotifier) processQueue() {
+	for event := n.nextItem(); event != nil && !n.fuse.IsBroken(); event = n.nextItem() {
+		if err := n.send(event); err != nil {
+			n.params.Logger.Warnw("failed to send webhook", err, "url", n.params.URL, "event", event.Event)
+			n.dropped.Inc()
```

Comment:
if `event.NumDropped` is nonzero `n.dropped` is off after failures

Reply:
good point.. i'll restore.